File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        if len(s) == 0:
            return 0
        
        best=1 
        boundary=0
        
        for i in range(1,len(s)):
            
            if s[i] in s[boundary:i]:
                
                boundary=s[boundary:i].index(s[i])+boundary+1
                
            best=max(best,i-boundary+1)
            
        return best

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        unordered_map<char, int> letters;
        
        int best = 0;
        int left = 0;
        int right = 0;
        
        while(right < s.size()){
            
            if(letters.find(s[right]) == letters.end()){
                letters[s[right]]++;
                best = max(best, (right - left + 1));
                right++;
            }
            else{
                letters.erase(s[left]);
                left++;
            }
            
        }
        return best;
  
    }
};